zcash_primitives/src/transaction/builder.rs

```diff
@@ -590,9 +591,32 @@ impl<'a, P: consensus::Parameters, R: RngCore> Builder<'a, P, R> {
     /// this function, and instead will generate a transaction that will be rejected by
     /// the network.
     pub fn build(
+        self,
+        consensus_branch_id: consensus::BranchId,
+        prover: &impl TxProver,
+    ) -> Result<(Transaction, TransactionMetadata), Error> {
+        self.build_with_progress_notifier(consensus_branch_id, prover, None)
+    }
+
+    /// Builds a transaction from the configured spends and outputs.
+    ///
+    /// Upon success, returns a tuple containing the final transaction, and the
+    /// [`TransactionMetadata`] generated during the build process.
+    ///
+    /// `consensus_branch_id` must be valid for the block height that this transaction is
+    /// targeting. An invalid `consensus_branch_id` will *not* result in an error from
+    /// this function, and instead will generate a transaction that will be rejected by
+    /// the network.
+    ///
+    /// `progress_notifier` sets the notifier channel, where progress of building the transaction is sent.
+    /// It sends an update after every Spend or Output is computed, and the `u32` sent represents
+    /// the total steps completed so far. It will eventually send number of spends + outputs.
```

Comment:
Notifying with a single `u32` doesn't enable the caller to determine the amount of progress, only that progress has been made. They could guess based on the number of times that `add_sapling_spend` and `add_sapling_output` were called, but a) that is only a lower bound on the actual number (since we pad outputs), and b) that isn't particularly usable (when we add Orchard, the caller has to remember to take them into account).

Instead, I propose that we use `Sender<(u32, Option<u32>)>`, or `Sender<builder::Progress>` where:
```rust
pub struct Progress {
    cur: u32,
    end: Option<u32>,
}

impl Progress {
    // Getters and/or rendering helper methods
}
```

This way, if the end-point is known (which in most cases for the transaction builder it will be), this information can be conveyed to the caller. If we use the `struct Progress` approach, then we could augment this in future with time-remaining estimation as well (instead of just percentage completeness).

Reply:
This is a good idea. Made this change